chainerrl/agents/dqfd.py

```diff
@@ -0,0 +1,679 @@
+
+import chainer
+from chainer import cuda
+import chainer.functions as F
+
+from chainerrl.agents import DoubleDQN
+
+from chainerrl.misc.batch_states import batch_states
+from chainerrl.replay_buffer import PrioritizedBuffer, PrioritizedReplayBuffer
+
+from future import standard_library
+standard_library.install_aliases()  # NOQA
+
+import numpy as np
+
+
+def compute_weighted_value_loss(y, t, weights,
+                                mask, clip_delta=True,
+                                batch_accumulator='mean'):
+    """Compute a loss for value prediction problem.
+
+    Args:
+        y (Variable or ndarray): Predicted values.
+        t (Variable or ndarray): Target values.
+        weights (ndarray): Weights for y, t.
+        mask (ndarray): Mask to use for loss calculation
+        clip_delta (bool): Use the Huber loss function if set True.
+        batch_accumulator (str): 'mean' will divide loss by batchsize
+    Returns:
+        (Variable) scalar loss
+    """
+    assert batch_accumulator in ('mean', 'sum')
+    y = F.reshape(y, (-1, 1))
+    t = F.reshape(t, (-1, 1))
+    if clip_delta:
+        losses = F.huber_loss(y, t, delta=1.0)
+    else:
+        losses = F.square(y - t) / 2
+    losses = F.reshape(losses, (-1,))
+    loss_sum = F.sum(losses * weights * mask)
+    if batch_accumulator == 'mean':
+        loss = loss_sum / max(mask.sum(), 1.0)
+    elif batch_accumulator == 'sum':
+        loss = loss_sum
+    return loss
+
+
+class PrioritizedDemoReplayBuffer(PrioritizedReplayBuffer):
+    """Replay buffer used in DQfD.
+
+    Differences from a normal PrioritizedReplayBuffer
+      Stores both both persistent demonstration data and normal self-play data
+      Stores both 1-step and n-step versions of every transition
+
+    Args:
+        capacity(int): Capacity of the buffer *excluding* expert demonstrations
+
+    Standard PER parameters:
+        alpha, beta0, betasteps, eps (float)
+        normalize_by_max (bool)
+    """
+
+    def __init__(self, capacity=None,
+                 alpha=0.6, beta0=0.4, betasteps=2e5, eps=0.01,
+                 normalize_by_max=True, error_min=0,
+                 error_max=2, num_steps=1):
+
+        PrioritizedReplayBuffer.__init__(self, capacity=capacity,
+                                         alpha=alpha, beta0=beta0,
+                                         betasteps=betasteps,
+                                         eps=eps,
+                                         normalize_by_max=normalize_by_max,
+                                         error_min=error_min,
+                                         error_max=error_max,
+                                         num_steps=num_steps)
+
+        self.memory = PrioritizedBuffer(capacity)
+        self.memory_demo = PrioritizedBuffer(None)
+
+    def weights_from_probabilities(self, probabilities, min_probability):
+        """Overwrite weights_from_probabilities to make beta increment explicit
```

Comment:
I think it would be better to not overwrite `weights_from_probabilities`, and instead modify the `PriorityWeightError`'s `weights_from_probabilities` to also take in a bool `update_beta`, which defaults to `True`. Then, in your `sample` function, you can take in a bool `pretraining`, which is then passed to `weights_to_probabilities` as `update_beta=pretraining`. That way, you can remove  the `update_beta` function, and in your `ReplayUpdater`, you can take in a flag `pretraining` that's passed through `update_if_necessary`, which can then be used when sampling transitions. Give this some thought. I think this drastically reduces duplicate code, sticks more with convention, and reduces likelihood of bugs.

Reply:
I tried doing this but since `weights_from_probabilities` is called twice- one for each buffer, I'll have to call one with `update_beta=True` and another with `update_beta=False` which is not so nice. I'll see if I can come up with a better way.